app/ts/context/BalanceContext.tsx

```diff
@@ -0,0 +1,94 @@
+import { Signal, useSignal } from '@preact/signals'
+import { BigNumber, ethers } from 'ethers'
+import { ComponentChildren, createContext } from 'preact'
+import { useContext } from 'preact/hooks'
+import { useAsyncState } from '../library/preact-utilities.js'
+import { assertsExternalProvider, assertUnreachable, isEthereumObservable } from '../library/utilities.js'
+
+type Balance =
+	| {
+			state: 'initial'
+			checkBalance: () => void
+	  }
+	| {
+			state: 'updating'
+			reset: () => void
+	  }
+	| {
+			state: 'updated'
+			balance: BigNumber
+			updated_at: Date
+	  }
+	| {
+			state: 'failed'
+			error: Error
+			reset: () => void
+	  }
+
+export type BalanceStore = Signal<Balance>
+export function createBalanceStore() {
+	const { value: query, waitFor, reset } = useAsyncState<BigNumber>()
+
+	const checkBalance = () => {
+		waitFor(async () => {
+			assertsExternalProvider(window.ethereum)
+			const provider = new ethers.providers.Web3Provider(window.ethereum)
+			const signer = provider.getSigner()
+			const address = await signer.getAddress()
+			return await provider.getBalance(address)
+		})
+	}
+
+	const balanceStoreDefaults = { state: 'initial' as const, checkBalance }
+	const balanceStore = useSignal<Balance>(balanceStoreDefaults)
+
+	const handleAccountChange = (newAccount: string[]) => {
+		if (ethers.utils.isAddress(newAccount[0])) {
```

Comment:
this can result in reading out of bounds if newAccount is empty

Reply:
replaced this with a `connect` function similar to how one would connect the first time. this approach had issues where a state change is expected by other conditions. Ex. the balance functions need to know the account changed without filtering the signal values as the state will have stayed 'connected' and only the address changed.